# Load necessary libraries
library(Seurat)
library(org.Hs.eg.db)
library(cowplot)
library(patchwork)
library(dplyr)
library(limma)
library(ggplot2)

# Source custom scripts
source("BETi-COVIDHeart/utilities/ggplotColors.R") 
source("BETi-COVIDHeart/utilities/normCounts.R")  

#---------Read in snRNAseq data of human cardiac organoids (hCO) generated by Mills et al. 2021 [EGAS00001005174]------------ 
# Load 10X data for Control sample
organoids.CTRL <- Read10X(data.dir = "./filtered_feature_bc_matrix/")
colnames(organoids.CTRL) <- paste(colnames(organoids.CTRL), "CTRL", sep = "_")
dim(organoids.CTRL)
#[1] 33538  5834

# Load 10X data for Stimulated sample 1 
organoids.STIM.1 <- Read10X(data.dir = "./filtered_feature_bc_matrix/")
colnames(organoids.STIM.1) <- paste(colnames(organoids.STIM.1), "STIM.1", sep = "_")
dim(organoids.STIM.1)
#[1] 33538  4500

# Load 10X data for Stimulated sample 2 
organoids.STIM.2 <- Read10X(data.dir = "./filtered_feature_bc_matrix/")
colnames(organoids.STIM.2) <- paste(colnames(organoids.STIM.2), "STIM.2", sep = "_")
dim(organoids.STIM.2)
#[1] 33538  2659

# Combine all samples
all.samples <- cbind(organoids.CTRL, organoids.STIM.1, organoids.STIM.2)
dim(all.samples)
#[1] 33538 12993

# --- Integration with snRNAseq of human fetal heart samples generated by Sim et al. 2020 [GSE156707]-----------
# Load existing combined hCOs Seurat object
DefaultAssay(hCOs.combined) <- "RNA"
hCOs_proliferative <- subset(hCOs.combined, subset = Celltype %in% c("Proliferative"))
#[1] 21915   670

# Load integrated fetal sample data
fetal.integrated <- readRDS(file = "./fetal-int.Rds")
load(file = "./fetalObjs.Rdata")
# Default resolution for clustering: 0.3
Idents(fetal.integrated) <- fetal.integrated$integrated_snn_res.0.3
DimPlot(fetal.integrated, reduction = "tsne", label = TRUE, label.size = 6) + NoLegend()
table(fetal.integrated$Celltype)
DefaultAssay(fetal.integrated) <- "RNA"
fetal_proliferative <- subset(fetal.integrated, subset = Celltype %in% c("Proliferative CM2"))
#[1] 3000  604
DefaultAssay(fetal_proliferative) <- "RNA"
#[1] 17723   604

# Find common genes between hCOs and fetal proliferative cells
index <- intersect(rownames(hCOs_proliferative), rownames(fetal_proliferative))
hCOs_proliferative <- hCOs_proliferative[index, ]
#[1] 17509   670
fetal_proliferative <- fetal_proliferative[index, ]
#[1] 17509   604

# Find common genes between combined hCOs and integrated fetal data
index <- intersect(rownames(hCOs.combined), rownames(fetal.integrated))
hCOs.combined <- hCOs.combined[index, ]
fetal.integrated <- fetal.integrated[index, ]
table(rownames(fetal.integrated) == rownames(hCOs.combined))

# Merge proliferative subsets
integrated_proliferative <- merge(hCOs_proliferative, y = fetal_proliferative, project = "Proliferative")
# Merge entire hCOs and fetal integrated datasets
integrated_proliferative <- merge(hCOs.combined, y = fetal.integrated, project = "fetal.hCOs")

# Add a broad cell type annotation
integrated_proliferative$Broad_celltype <- "Proliferative CMs"
table(integrated_proliferative$Broad_celltype)
# Proliferative CMs
#        1274

# Split merged object by biorep
integrated_proliferative.list <- SplitObject(integrated_proliferative, split.by = "biorep")
min(sapply(integrated_proliferative.list, ncol))

# Apply SCTransform to each list element
for (i in 1:length(integrated_proliferative.list)) {
  integrated_proliferative.list[[i]] <- SCTransform(integrated_proliferative.list[[i]], verbose = FALSE)
}

# Find integration anchors
cardio.anchors <- FindIntegrationAnchors(object.list = integrated_proliferative.list, dims = 1:30,
                                         anchor.features = 3000, k.filter = 1872)
# Integrate the data
cardio.integrated <- IntegrateData(anchorset = cardio.anchors, dims = 1:30)

# Seurat workflow
DefaultAssay(object = cardio.integrated) <- "integrated"
cardio.integrated <- ScaleData(cardio.integrated, verbose = FALSE)
cardio.integrated <- RunPCA(cardio.integrated, npcs = 50, verbose = FALSE)
ElbowPlot(cardio.integrated, ndims = 50)
cardio.integrated <- FindNeighbors(cardio.integrated, dims = 1:20)
cardio.integrated <- FindClusters(cardio.integrated, resolution = 0.1)
table(Idents(cardio.integrated))
set.seed(10)
cardio.integrated <- RunTSNE(cardio.integrated, reduction = "pca", dims = 1:20)

# Save and load integrated data
save(cardio.integrated, file = "./fetal_hCOs.Rdata")
load(file = "./fetal_hCOs.Rdata")

# Visualize integrated data
DimPlot(cardio.integrated, reduction = "tsne", label = TRUE, label.size = 6, pt.size = 0.5) + NoLegend()
DimPlot(cardio.integrated, reduction = "tsne", split.by = "orig.ident", label = TRUE)

# Subset data based on cell type (various subsets explored)
cardio_nonProliferative <- subset(cardio.integrated,
                                  subset = Celltype %in% c("Cardiomyocyte (conduction)",
                                                           "Cardiomyocytes (Female)",
                                                           "Cardiomyocytes (Male)",
                                                           "Cardiomyocytes1",
                                                           "Cardiomyocytes2"))
cardio_nonProliferative <- subset(cardio.integrated,
                                  subset = Celltype %in% c("Proliferative CM2",
                                                           "Proliferative CM1",
                                                           "Proliferative CM (Male)"))
cardio_nonProliferative <- subset(cardio.integrated, subset = Celltype == "ProliferativehCOs")
cardio_nonProliferative <- subset(cardio.integrated,
                                  subset = Celltype %in% c("Endothelial cells",
                                                           "Endothelial cells/ Immune?",
                                                           "Lymphatic endothelial cells",
                                                           "Vascular (endothelial?) (Female)"))
cardio_nonProliferative <- subset(cardio.integrated,
                                  subset = Celltype %in% c("Fibroblast (Female?)",
                                                           "Fibroblast1",
                                                           "Fibroblast2"))
cardio_nonProliferative <- subset(cardio.integrated, subset = Celltype %in% c("FibroblasthCOs"))
cardio_nonProliferative <- subset(cardio.integrated, subset = Celltype %in% c("CardiomyocyteshCOs"))

# Relabeling cells based on subsetting
index <- match(colnames(cardio_nonProliferative), colnames(cardio.integrated))
table(cardio.integrated$integrated_snn_res.0.1[index])
cardio.integrated$Celltype[index] <- "hCOs_Cardiomyocytes"

# Create data frame for cluster proportions
clusters <- rep(0:9, 14)
condition <- rep(c("hCOs_Proliferative CMs", "Proliferative CMs", "hCOs_Cardiomyocytes",
                   "Cardiomyocytes", "hCOs_Fibroblasts", "Fibroblasts", "Endothelials",
                   "Epicardial-like", "Erythroid", "Macrophage", "Smooth muscle cells",
                   "Neurons", "Pericyte", "Mast cells"), each = 10)
value <- rep(0, 140)
data1 <- data.frame(clusters, condition, value)
group <- c("hCOs_Proliferative CMs", "Proliferative CMs", "hCOs_Cardiomyocytes",
           "Cardiomyocytes", "hCOs_Fibroblasts", "Fibroblasts", "Endothelials",
           "Epicardial-like", "Erythroid", "Macrophage", "Smooth muscle cells",
           "Neurons", "Pericyte", "Mast cells")

# Calculate cluster proportions for each cell type
j = 1
for (i in 1:14) {
  type <- subset(cardio.integrated, subset = Celltype == group[i])
  index <- match(colnames(type), colnames(cardio.integrated))
  tab <- table(cardio.integrated$integrated_snn_res.0.1[index])
  # data1$value[j:(j+12)] <- tab
  data1$value[j:(j + 9)] <- round(tab / table(cardio.integrated$integrated_snn_res.0.1) * 100, digits = 1)
  j = j + 10
}

# Test subsetting (Pericyte example)
table(cardio.integrated$integrated_snn_res.0.1)
cardio_nonProliferative <- subset(cardio.integrated, subset = Celltype %in% c("Pericyte"))
index <- match(colnames(cardio_nonProliferative), colnames(cardio.integrated))
table(cardio.integrated$integrated_snn_res.0.1[index])

# Plot cluster proportions
library(ggplot2)
library("RColorBrewer")
ggplot(data1, aes(fill = factor(condition,
                                levels = c("hCOs_Proliferative CMs", "Proliferative CMs",
                                           "hCOs_Cardiomyocytes", "Cardiomyocytes",
                                           "hCOs_Fibroblasts", "Fibroblasts", "Endothelials",
                                           "Epicardial-like", "Erythroid", "Macrophage",
                                           "Smooth muscle cells", "Neurons", "Pericyte",
                                           "Mast cells")),
                  y = value, x = clusters)) +
  geom_bar(position = "stack", stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 0.5)) +
  scale_x_continuous("Cluster", labels = as.character(clusters), breaks = clusters) +
  scale_fill_manual(values = c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C", "#FB9A99",
                               "#E31A1C", "#FDBF6F", "#FF7F00", "#CAB2D6", "#6A3D9A",
                               "#FFFF99", "#B15928", "#D9D9D9", "#FFED6F"))

aspect_ratio <- 2.5
ggsave(plot = p, width = 7 * aspect_ratio, height = 7, dpi = 300,
       filename = "./fetal_hCOs_res0.1.jpg")

# Barplot of cell type proportions per cluster
par(mfrow = c(1, 1))
par(mar = c(4, 4, 2, 2))
tab <- table(Idents(cardio.integrated), cardio.integrated$Celltype)
barplot(t(tab / rowSums(tab)), beside = TRUE, col = ggplotColors(23), legend = TRUE)

# Barplot of sample proportions per cluster
par(mfrow = c(1, 1))
par(mar = c(4, 4, 2, 2))
tab <- table(Idents(cardio.integrated), cardio.integrated$orig.ident)
barplot(t(tab / rowSums(tab)), beside = TRUE, col = ggplotColors(2))
legend("topleft", legend = colnames(tab), fill = ggplotColors(2))

# Visualize biorep distribution
cardio.integrated$biorep <- factor(cardio.integrated$biorep,
                                   levels = c("f1", "f2", "f3", "CTRL", "STIM.1", "STIM.2"))
DimPlot(cardio.integrated, reduction = "tsne", group.by = "biorep")

# Barplot of biorep proportions per cluster
par(mfrow = c(1, 1))
par(mar = c(4, 4, 2, 2))
tab <- table(Idents(cardio.integrated), cardio.integrated$biorep)
barplot(t(tab / rowSums(tab)), beside = TRUE, col = ggplotColors(6), legend = TRUE, ylim = c(0, 0.5))
title(main = "Percentage of Cells in Each Cluster")

# Barplot of original sample proportions per cluster (repeated)
par(mfrow = c(1, 1))
par(mar = c(4, 4, 2, 2))
tab <- table(Idents(cardio.integrated), cardio.integrated$orig.ident)
barplot(t(tab / rowSums(tab)), beside = TRUE, col = ggplotColors(2), ylim = c(0, 0.8))
legend("topleft", legend = colnames(tab), fill = ggplotColors(2))
title(main = "Percentage of Cells in Each Cluster")

# Differential expression analysis with edgeR
DefaultAssay(cardio.integrated) <- "RNA"
Idents(cardio.integrated) <- cardio.integrated$integrated_snn_res.0.1
all <- cardio.integrated@assays$RNA@counts
numzero.genes <- rowSums(all == 0)
table(numzero.genes > (ncol(all) - 20))
keep.genes <- numzero.genes < (ncol(all) - 20)
table(keep.genes)
all <- all[keep.genes, ]
dim(all)
library(edgeR)
y.cardio <- DGEList(all)
logcounts <- normCounts(y.cardio, log = TRUE, prior.count = 0.5)

# Define design matrix for differential expression
maxclust <- length(levels(Idents(cardio.integrated))) - 1
grp <- paste("c", Idents(cardio.integrated), sep = "")
grp <- factor(grp, levels = paste("c", 0:maxclust, sep = ""))
design <- model.matrix(~0 + grp + cardio.integrated$biorep)
colnames(design)[1:(maxclust + 1)] <- levels(grp)

# Create contrast matrix for pairwise comparisons between clusters
mycont <- matrix(0, ncol = length(levels(grp)), nrow = length(levels(grp)))
colnames(mycont) <- levels(grp)
diag(mycont) <- 1
mycont[upper.tri(mycont)] <- -1 / (length(levels(factor(grp))) - 1)
mycont[lower.tri(mycont)] <- -1 / (length(levels(factor(grp))) - 1)

# Fill out remaining rows with 0s for biorep effects
zero.rows <- matrix(0, ncol = length(levels(grp)),
                    nrow = (ncol(design) - length(levels(Idents(cardio.integrated)))))
test <- rbind(mycont, zero.rows)

# Fit linear model and apply contrasts
fit <- lmFit(logcounts, design)
fit.cont <- contrasts.fit(fit, contrasts = mycont) # Apply contrasts to the linear model fit
fit.cont <- eBayes(fit.cont, trend = TRUE, robust = TRUE) # Empirical Bayes moderation of standard errors

# Annotate genes with various identifiers
ann <- AnnotationDbi::select(org.Hs.eg.db,
                             keys = rownames(fit.cont), # Use rownames from fit.cont, assuming it contains gene symbols
                             columns = c("SYMBOL", "ENTREZID", "ENSEMBL", "GENENAME", "CHR"),
                             keytype = "SYMBOL")
fit.cont$genes <- ann # Add gene annotations to the fit.cont object

# Summarize the differential expression results
summary(decideTests(fit.cont))

# Apply a log fold change threshold
treat <- treat(fit.cont, lfc = 0.5)

# Decide differential expression with the new threshold
dt <- decideTests(treat)
summary(dt)

# Plot MD plots
par(mfrow = c(1, 3))
for (i in 1:ncol(mycont)) {
  plotMD(treat, coef = i, status = dt[, i], hl.cex = 0.5)
  abline(h = 0, col = colours()[226])
  lines(lowess(treat$Amean, treat$coefficients[, i]), lwd = 1.5, col = 4)
}

# Extract contrast names
contnames <- colnames(mycont)

# Get the top significant genes
topsig <- topTreat(treat, coef = 2, n = Inf)

# Filter for significant and upregulated genes
up_topsig <- topsig %>% filter(adj.P.Val < 0.05 & logFC > 0.5)

# Write the filtered results to a CSV file.  
write.csv(up_topsig, file = "./Cluster-2.csv") 

# Loop through contrasts and write top genes to CSV files.
for (i in 1:length(contnames)) {
  topsig <- topTreat(treat, coef = i, n = Inf)
  write.csv(topsig, file = paste0("./Cluster-", contnames[i], ".csv")) 
}

# Perform GO analysis
GO_Cluster <- topGO(goana(de = up_topsig$ENTREZID, species = "Hs"), number = Inf)

# Prepare plot parameters
par(las = 2) # make label text perpendicular to axis
par(mar = c(5, 15, 4, 2)) # increase y-axis margin.

# Filter GO terms
GO_0 <- GO_Cluster %>% filter(GO_Cluster$Ont == "BP" & GO_Cluster$P.DE < 0.05)

# Select relevant columns
GO_0_selected <- GO_0 %>% dplyr::select(c(Term, P.DE, DE))

# Further filter based on gene count
GO_0_selected <- GO_0_selected %>% filter(DE > 5)

# Write the GO analysis results to a CSV file. 
write.csv(GO_0_selected, file = "./GOProliferativeCMs_c11.csv") 

# Create a barplot of the top GO terms
bp <- barplot(GO_0_selected$DE[10:1], horiz = TRUE, names.arg = GO_0_selected$Term[10:1], cex.names = 0.8, cex.axis = 0.8, xlim = c(0, 140))
title(main = "Top 10 significant Biological Processes in Proliferative", xlab = "#Genes in DE list")

# Load pre-existing R data objects. 
load("human_c2_v5p2.rdata") 
load("human_c5_v5p2.rdata")

# Map gene IDs
c2.id <- ids2indices(Hs.c2, treat$genes$ENTREZID) # KEGG & REACTOME
c5.id <- ids2indices(Hs.c5, treat$genes$ENTREZID) # GO

# Filter for REACTOME pathways
reactome.id <- c2.id[grep("REACTOME", names(c2.id))]

# Perform pathway enrichment analysis
c2.c0 <- cameraPR(treat$t[, 1], c2.id)
reactome.c0 <- cameraPR(treat$t[, 1], reactome.id)
go.c0 <- cameraPR(treat$t[, 1], c5.id)

# Filter GO results
go.c0 <- go.c0 %>% filter(Direction == "Up" & FDR < 0.05 & NGenes > 5)

# Example gene match
m <- c5.id[["GO_CELLULAR_METABOLIC_PROCESS"]]
match('2293', ann$ENTREZID)

# Loop through contrasts and write pathway analysis results to CSV files.  .
for (i in 1:length(contnames)) {
  write.csv(cameraPR(treat$t[, i], c2.id), file = paste0("./c2-", contnames[i], ".csv")) 
  write.csv(cameraPR(treat$t[, i], reactome.id), file = paste0("./reactome-", contnames[i], ".csv")) 
  write.csv(cameraPR(treat$t[, i], c5.id), file = paste0("./go-", contnames[i], ".csv"))
}

# Define species and database
species <- "Hs"
DB <- paste("org", species, "eg", "db", sep = ".")
require(DB, character.only = TRUE)
GO2ALLEGS <- paste("org", species, "egGO2ALLEGS", sep = ".")
EG.GO <- AnnotationDbi::toTable(get(GO2ALLEGS))

# Remove duplicated entries
d <- duplicated(EG.GO[, c("gene_id", "go_id", "Ontology")])
EG.GO <- EG.GO[!d, ]

# Split genes by GO ID/Ontology
de.by.go <- split(EG.GO$gene_id, paste(EG.GO$go_id, EG.GO$Ontology, sep = "."))
de = up_topsig$ENTREZID
de.by.go <- lapply(de.by.go, FUN = function(x) {
  x[x %in% de]
})

# Select pathways
GO <- paste(rownames(GO_0_selected), "BP", sep = ".")
pathways <- de.by.go[GO]
m <- de.by.go[['GO:0015031.BP']]

# Match gene symbols
index <- match(m, up_topsig$ENTREZID)
up_topsig[index, 1]

# Print gene symbols for each pathway
for (i in 1:length(pathways)) {
  print(ann$SYMBOL[match(pathways[[i]][1:length(pathways[[i]])], ann$ENTREZID)])
}
